compiler/hash-typecheck/src/ops/typing.rs

```diff
@@ -330,34 +380,63 @@ impl<'gs, 'ls, 'cd, 's> Typer<'gs, 'ls, 'cd, 's> {
     /// subject.
     pub(crate) fn get_term_of_pat(&mut self, pat_id: PatId) -> TcResult<TermId> {
         let pat = self.reader().get_pat(pat_id).clone();
+
         let ty_of_pat = match pat {
             Pat::Mod(_) | Pat::Ignore | Pat::Binding(_) => {
                 // We don't know this; it depends on the subject:
                 Ok(self.builder().create_unresolved_term())
             }
+            Pat::Const(ConstPat { term, .. }) => Ok(term),
+            Pat::Access(AccessPat { subject, property }) => {
+                let subject_id = self.get_term_of_pat(subject)?;
+
+                Ok(self.builder().create_access(subject_id, property, AccessOp::Namespace))
+            }
             Pat::Lit(lit_term) => {
                 // The term of a literal pattern is the literal (lol):
                 Ok(lit_term)
             }
             Pat::Tuple(tuple_pat) => {
                 // For each parameter, get its type, and then create a tuple
                 // type:
-                let params_id = self.infer_args_of_pat_params(tuple_pat)?;
+                let params_id = self.infer_args_of_pat_args(tuple_pat)?;
+                Ok(self.builder().create_tuple_lit_term(params_id))
+            }
+            Pat::Constructor(constructor_pat) => {
+                // We have params to apply, so we need to create an FnCall
+                let args_id = self.infer_args_of_pat_args(constructor_pat.args)?;
+                Ok(self.builder().create_constructed_term(constructor_pat.subject, args_id))
+            }
+            Pat::List(ListPat { term, .. }) => {
+                // We want to create a `List<T = term>` as the type of the pattern
+                let list_inner_ty = self.core_defs().list_ty_fn;
                 let builder = self.builder();
-                Ok(builder.create_tuple_lit_term(params_id))
+
+                let list_ty = builder.create_app_ty_fn_term(
+                    list_inner_ty,
+                    builder.create_args([builder.create_arg("T", term)], ParamOrigin::TyFn),
+                );
+
+                Ok(builder.create_rt_term(list_ty))
```

Comment:
@@Future: use a list literal term instead

Reply:
there is/will be a literal list term?

Reply:
Ideally, there should be, i dont see why not..

Reply:
Not something to worry about now tho

Reply:
In which case `x := [1, 2, 3]; x[0]` would be eta reduced to `1` during TC